onnxruntime/core/providers/cuda/tensor/split_impl.cu

```diff
@@ -193,82 +212,53 @@ __device__ __forceinline__ void _block_copy_data(T* out, const T* input, int siz
+  auto input_v = reinterpret_cast<size_t>(input_data);
+  auto output_v0 = reinterpret_cast<size_t>(output_data0);
+  auto output_v1 = reinterpret_cast<size_t>(output_data1);
+  auto output_v2 = reinterpret_cast<size_t>(output_data2);
+  auto size0_in_byte = size0 * element_size;
+  auto size1_in_byte = size1 * element_size;
+  auto size2_in_byte = size2 * element_size;
+
+  auto VEC_SIZE = std::min(select(size0_in_byte), std::min(select(size1_in_byte), select(size2_in_byte)));
+  auto min_output_vec_size = std::min(select(output_v0), std::min(select(output_v1), select(output_v2)));
+  VEC_SIZE = std::min(VEC_SIZE, std::min(select(input_v), min_output_vec_size));
+
+  // determine threads based on the size of the output
+  auto threadsPerBlock = kNumThreadsPerBlock;
+  if ((inner_size_in_byte / VEC_SIZE) < 128) {
+    // use less threads when the size is small
+    threadsPerBlock = 128;
   }
-  CUDA_LONG inner_size = static_cast<CUDA_LONG>(input_shape[input_shape.size() - 1]);
 
-  int blocksPerGrid = outer_size * 3;
-
-  switch (element_size) {
-#define CASE_ELEMENT_TYPE(type)                                                                 \
-  case sizeof(type): {                                                                          \
-    _Split3InnerKernel<<<blocksPerGrid, kNumThreadsPerBlock, 0, stream>>>(size0, size1, size2,   \
-        reinterpret_cast<const ToCudaType<type>::MappedType*>(input_data),                       \
-        reinterpret_cast<ToCudaType<type>::MappedType*>(output_data0),                           \
-        reinterpret_cast<ToCudaType<type>::MappedType*>(output_data1),                           \
-        reinterpret_cast<ToCudaType<type>::MappedType*>(output_data2), outer_size, inner_size);  \
-  } break
-    CASE_ELEMENT_TYPE(int8_t);
-    CASE_ELEMENT_TYPE(int16_t);
-    CASE_ELEMENT_TYPE(int32_t);
-    CASE_ELEMENT_TYPE(int64_t);
-#undef CASE_ELEMENT_TYPE
+  switch (VEC_SIZE) {
+#define CASE_ELEMENT_TYPE(type)                                                                       \
+    _Split3InnerKernel<type><<<outer_size, threadsPerBlock, 0, stream>>>(                             \
+                                                            size0_in_byte,                            \
+                                                            size1_in_byte,                            \
+                                                            size2_in_byte,                             \
+                                                            input_data,        \
+                                                            output_data0,      \
+                                                            output_data1,      \
+                                                            output_data2,      \
+                                                            inner_size_in_byte)
+    case 16:
+      CASE_ELEMENT_TYPE(int4);
+      break;
+    case 8:
+      CASE_ELEMENT_TYPE(int2);
```

Comment:
I prefer data types int8_t, int16_t, int32_t, int64_t, int4. That's more common than int1, int2.